### ISSUE TITLE
[YSQL] 'Query error: Cannot decode value type kGroupEnd from the key encoding format: "!"' due to DISTINCT clause

### ISSUE THREAD

**Issue by mislam-yb** [yugabyte/yugabyte-db#16038]
Jira Link: [DB-5437](https://yugabyte.atlassian.net/browse/DB-5437)
### Description

This error comes up specifically from a Hasura generated query, and is related to distinct clauses.

Minimal reproduction:

```
CREATE TABLE t1 (k int primary key);

SELECT
	COALESCE(
	  json_agg(
	    row_to_json(info)
	  ),
	  '[]' :: JSON
	)
	FROM (
	SELECT
	tc.table_schema,
	tc.table_name,
	tc.constraint_name,
	json_agg(constraint_column_usage.column_name) AS columns
	FROM
	information_schema.table_constraints tc
	JOIN (
	  SELECT
	    x.tblschema AS table_schema,
	    x.tblname AS table_name,
	    x.colname AS column_name,
	    x.cstrname AS constraint_name
	  FROM ( SELECT DISTINCT
	      nr.nspname,
	      r.relname,
	      a.attname,
	      c.conname
	    FROM
	      pg_namespace nr,
	      pg_class r,
	      pg_attribute a,
	      pg_depend d,
	      pg_namespace nc,
	      pg_constraint c
	    WHERE
	      nr.oid = r.relnamespace
	      AND r.oid = a.attrelid
	      AND d.refclassid = 'pg_class'::regclass::oid
	      AND d.refobjid = r.oid
	      AND d.refobjsubid = a.attnum
	      AND d.classid = 'pg_constraint'::regclass::oid
	      AND d.objid = c.oid
	      AND c.connamespace = nc.oid
	      AND c.contype = 'c'::"char"
	      AND(r.relkind = ANY (ARRAY ['r'::"char", 'p'::"char"]))
	      AND NOT a.attisdropped
	    UNION ALL
	    SELECT
	      nr.nspname,
	      r.relname,
	      a.attname,
	      c.conname
	    FROM
	      pg_namespace nr,
	      pg_class r,
	      pg_attribute a,
	      pg_namespace nc,
	      pg_constraint c
	    WHERE
	      nr.oid = r.relnamespace
	      AND r.oid = a.attrelid
	      AND nc.oid = c.connamespace
	      AND r.oid = CASE c.contype
	      WHEN 'f'::"char" THEN
	        c.confrelid
	      ELSE
	        c.conrelid
	      END
	      AND(a.attnum = ANY (
	          CASE c.contype
	          WHEN 'f'::"char" THEN
	            c.confkey
	          ELSE
	            c.conkey
	          END))
	      AND NOT a.attisdropped
	      AND(c.contype = ANY (ARRAY ['p'::"char", 'u'::"char", 'f'::"char"]))
	      AND(r.relkind = ANY (ARRAY ['r'::"char", 'p'::"char"]))) x (tblschema, tblname, colname, cstrname)) constraint_column_usage ON tc.constraint_name::text = constraint_column_usage.constraint_name::text
	  AND tc.table_schema::text = constraint_column_usage.table_schema::text
	  AND tc.table_name::text = constraint_column_usage.table_name::text
	where ((tc.table_schema='public'))
	  AND tc.constraint_type::text = 'PRIMARY KEY'::text
	GROUP BY
	  tc.table_schema, tc.table_name, tc.constraint_name) as info;
```

The symptoms are the same as [this issue](https://github.com/yugabyte/yugabyte-db/issues/12422), however we believe the causes are different as this one is due to the disctinct clause.

[DB-5437]: https://yugabyte.atlassian.net/browse/DB-5437?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by bmatican** [comment]
cc @karan-yb

**Comment by tanujnay112** [comment]
It appears that the `prefix_length_` here is 23/24 (I forgot which one) when the table only has two primary keys so it is pertinent to the DISTINCT pushdown implementation.

**Comment by tanujnay112** [comment]
Closed by 977491e7d6426f84dc00723b01bac2614e548f1e
